### ISSUE TITLE
[ECS] : ECS Support for IPv6

### ISSUE THREAD

**Issue by vibhav-ag** [aws/containers-roadmap#1340]
Hi all,

Looking to get input on your IPv6 requirements as we develop our strategy here. IPv6 for ECS is a broad area and your feedback will help us prioritize our feature delivery to support the most important use cases.

Some topics that would be especially useful to get clarity on:

- ECS already supports dual stack (IPv4+IPv6) for Host and AWSVPC Networking modes. Do you need this to be extended to Bridge Networking mode?
- Do you need IPv6-only (IPv4 disallowed) access?
- What type of VPC resources do you want to access over IPv6?
- Are you planning to use IPv6 only within your VPC(s), or are you also planning to connect your Tasks to IPv6 internet?
- Do you require image pulls from ECR over IPv6?
- Do you plan to use ECS Service Discovery and/or AWS App Mesh with your ECS workloads with IPv6?

Anything else that is important to you!

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

**Which service(s) is this request for?**
ECS, Fargate

**Comment by jfrlb** [comment]
1. Bridge mode would be extremely interesting, as we are trying to run many application instances on "own" (and separate) IPv6 addresses within the same EC2 instance. We have rather small containers that need something like 0.25 vCPU and 200MB RAM, so we would like to run several of those on small t3-instances, but that does not work with awsvpc mode due to the ENI limits (and not trunking on t-instances). Currently we are working around it with docker host mode, but I'd rather say that's really just a work around with associated disadvantages (e.g. container has to run as root).
2. We are absolutely lacking the ability to "fix" the IPv6 addresses used by the containers. Our (client) application needs to contact our backend behind static IPv6 addresses but currently ECS only allows to specify the /64 prefix; but we'd rather need something like an "elastic IP" for our backend service. It would be a great starting point already, if a "secondary" IPv6 could be configured to the EC2's ENI. Through docker host mode we could use that as well in our containers. It is already possible to configure this on the linux machine, but it looks like the additional address is not considered in the VPC routing and the VPC thus does not forward any packets to such addresses today.
3. Network load balancer does not support UDP + IPv6 combination!! And no IPv6 for internal load balancers at all.
4. We would need IPv6 multicast support within VPC networking (we are receiving traffic into the VPC through a VPN endpoint and would like to forward received multicast packets to a number of ECS containers who are listening for these)

**Comment by jfrlb** [comment]
5. Especially for development and test stages it would be beneficial if AWS/VPC would also support ULA addressing (RFC4193) for VPC traffic. This would allow to create setups with specified, static and independent IPv6 addresses that you can rely on.

**Comment by zoltanmajo** [comment]
For our setup currently running in production we'd need IPv6 support for tasks launched by ECS services. That way we could access services outside of the VPC using IPv6.

So from the points mentioned in the description the following matches our requirements the best:
"Are you planning to use IPv6 only within your VPC(s), or are you also planning to connect your Tasks to IPv6 internet?"

The answer is yes. Many thanks in advance.

**Comment by leezu** [comment]
> Do you need IPv6-only (IPv4 disallowed) access?

Definitely. That's because AWS has bad support for IPv4 addresses on multi-network interface instance types such as p4d.
Multi-NIC instances do not support association of public IPv4 addresses at startup (they only support association of private IPv4 addresses and IPv6 addresses). Thus multi-NIC EC2 instances currently can't join the ECS clusters until an EIP is manually assigned or a NAT Gateway is configured in the subnet. NAT Gateway adds extra costs to customers and automatic EIP assignment is hard to configure for customers (requires implementation of Lambda functions based on SNS events of p4 instance startup).

Once ECS supports IPv6-only (IPv4 disallowed) access, ECS will work out-of-the box for multi-NIC instances.

**Comment by mogadanez** [comment]
>ECS already supports dual stack (IPv4+IPv6) for Host and AWSVPC Networking modes.

but there is a lack of documentation on how to properly connect it to ALB.

ECS works with **instance** type of target group, 
but in loadbalancer  setting I cant create **instance** type with ipv6 selection

**Comment by tlaverdure** [comment]
> Do you require image pulls from ECR over IPv6?

Yes, this would great for IPv6 egress only workloads.

**Comment by martin-pronestor** [comment]
I was experimenting setting up a VPC using dualstack and having two types of subnets (two public and one private) for a simple ECS hosted web application with an elastic load balancer in front. The images for the web application were hosted on ECR and I wanted to see if I could avoid having a NAT gateway (and elastic IPv4) associated to the private subnet and only use IPv6 and an egress-only gateway.

I was a little disappointed to find out that images couldn't be pulled from ECR to a running ECS with IPv6. It didn't even work if I used a different image repository.

`Do you require image pulls from ECR over IPv6?`

Yes, please. Hopefully this could fix our issue and we could save the cost associated with using a NAT gateway and Elastic IP.

**Comment by martin-pronestor** [comment]
> I don't get it. If you use and external repo like Docker Hub, egress-only gateway with IP6 not works?
From the error message that I received, when the ECS task was provisioning, it said it got a TCP timeout using the IPv4 address of the repository, so I assume the agent responsible for pulling images doesn't know how to use the IPv6 version of the repository endpoint. In my tests I tried pulling from ECR and from GitHub private container registry (https://docs.github.com/en/packages/working-with-a-github-packages-registry/working-with-the-container-registry).

> Why you need and Elastic IP?. Based on your requirements maybe VPC endpoint is cheaper than NAT gateway
From what I've been told you have to have an elastic IP in order to use a managed NAT gateway. (https://stackoverflow.com/questions/43094786/why-does-a-aws-nat-gateway-require-an-elasticip)
I tried setting up VPC endpoint but ultimately found that it ends up being almost as expensive as the NAT gateway because you need to add S3, ECR-DKR, ECR-API, and possibly CloudWatch (https://docs.aws.amazon.com/AmazonECR/latest/userguide/vpc-endpoints.html). It is also way harder to set up and debug :(

**Comment by gregwebs** [comment]
Can we create separate issues to support an IPv6 egress only gateway? A lot of the comments here are in support of this and I think there are too many other things being asked about on this issue.

Personally I would really like IPv6 egress only because it is a more secure architecture to be able to block ingress at the gateway level.

**Comment by relgames** [comment]
Hi, it would be nice to support IPv6 addresses to register ECS tasks in ALB target group (awsvpc mode).
Right now, when I tried to enable IPv6 in the target group, ECS is trying to use ipv4 to register in the ipv6 target group:
```
service test-EcsService-fYJDJS3mLw5J failed to register targets in target-group test-tg  with (error The IP address '172.17.2.196' is not a valid IPv6 address)
```
The task itself has IPv6:
```
Private IP
172.17.2.196

IPv6 address
2a05:d018:1cee:b902:bd1e:854e:44a4:1f78
```

**Comment by mbtran1** [comment]
> Hi @vibhav-ag, as of today it's possible to launch fargate tasks in a dual stack subnet (ipv4 + ipv6), are there any plans to support ipv6-only subnets in the future?

hi @julienp-b2wise and team, good timing of your inquiry. We (ECS) have plans for expanded IPv6 support, where ECS will deliver IPv6 support for its tasks and services on all compute platforms in dual stack (IPv4+IPv6) and IPv6-only subnets, with load balancer integration, image pulls, Domain Name System (DNS), and Task Metadata Service (TDMS) over IPv6.

Couple questions -- Please share your requirements in terms of ECR public IPv6 endpoint support and ECR PrivateLink IPv6 endpoint support for production deployments. Is there a hard requirement for ECR public IPv6 endpoint support for production deployments in 2023? Thanks.

**Comment by ngoyal16** [comment]
> Hi, it would be nice to support IPv6 addresses to register ECS tasks in ALB target group (awsvpc mode). Right now, when I tried to enable IPv6 in the target group, ECS is trying to use ipv4 to register in the ipv6 target group:
> 
> ```
> service test-EcsService-fYJDJS3mLw5J failed to register targets in target-group test-tg  with (error The IP address '172.17.2.196' is not a valid IPv6 address)
> ```
> 
> The task itself has IPv6:
> 
> ```
> Private IP
> 172.17.2.196
> 
> IPv6 address
> 2a05:d018:1cee:b902:bd1e:854e:44a4:1f78
> ```

We are also looking solution for same. @relgames  have you found any alternative.

**Comment by grahamschuckman** [comment]
Adding in the federal perspective here. Our agency client is responding to the White House OMB Mandate requiring that federal agencies begin operating in IPv6-only modes. See Page 3, Section 4a: https://www.whitehouse.gov/wp-content/uploads/2020/11/M-21-07.pdf

As such, we're looking to migrate our ECS workloads into an IPv6-only mode of operation.

**Comment by CGarces** [comment]
If we can't connect to ECR with IP6, there will be some impact con bills in 2024.
Run big Fargate clusters (each one with a public IPv4) will be more expensive.
https://aws.amazon.com/blogs/aws/new-aws-public-ipv4-address-charge-public-ip-insights/

**Comment by IllarionovDimitri** [comment]
Ran today onto the same issue. "IPv6 only" configuration is not yet supported by AWS, so the public IPv4 has to be assigned to the tasks running in public subnets. If there is no solution until Feb. 2024 event attaching IPv6 will not prevent the costs increase for public IPv4

**Comment by tophercullen** [comment]
> > Hi, it would be nice to support IPv6 addresses to register ECS tasks in ALB target group (awsvpc mode). Right now, when I tried to enable IPv6 in the target group, ECS is trying to use ipv4 to register in the ipv6 target group:
> > ```
> > service test-EcsService-fYJDJS3mLw5J failed to register targets in target-group test-tg  with (error The IP address '172.17.2.196' is not a valid IPv6 address)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > The task itself has IPv6:
> > ```
> > Private IP
> > 172.17.2.196
> > 
> > IPv6 address
> > 2a05:d018:1cee:b902:bd1e:854e:44a4:1f78
> > ```
> 
> We are also looking solution for same. @relgames have you found any alternative.

@ngSame issue here with running Fargate tasks. It had the added benefit of causing all sorts of internal breakage with AWS. Tasks and services stuck in various defunct states for hours despite deleting.

**Comment by ngoyal16** [comment]
@tophercullen what we have did is we have added a event bridge listener for task creation and invoke lammda and adding the ipv6 as target to nlb... That what we are doing as of now..but if AWS team can this own our behalf would be much better solution

**Comment by majkel-hyggio** [comment]
Guys, isn't that the anwser https://docs.aws.amazon.com/vpc/latest/userguide/nat-gateway-nat64-dns64.html ? 
If I get it right, that switch should enable EC2 ipv6 only instances to connect to ECR.

**Comment by tophercullen** [comment]
@majkel-hyggio You will incur additional costs in both NAT gateway hours and data transfer. NAT data transfer costs are not cheap, and if you don't setup a NAT per subnet (multiply NAT gateway hour charges), you'll also incur regional data transfer costs. Depending on the various elements (task durations, counts, image sizes etc.) it may actually be cheaper to just to allocate ipv4 addresses.

So no, this isn't really an acceptable answer to AWS ECR not having an IPv6 endpoint available.

**Comment by jnawk** [comment]
Given there's a charge for public IPv4 addresses coming in February, it's pretty important that there's a status-quo (ie free) alternative available, lest the change be viewed as purely an opportunistic money grab under the guise of IP address exhaustion.

**Comment by leonardlin** [comment]
I was caught off-guard by this too.

In preparation to ipv6 migration push (charge on public ipv4), I tried to move as much as possible to ipv6.

Turns out ec2 instances without public ipv4 can't join ecs-clusters.

ECS do not support IPv6 EC2 instances, even if they have a private IPv4

See also here: https://stackoverflow.com/questions/38599701/register-ec2-instance-to-ecs-cluster-without-public-ip

**Comment by ajpauwels** [comment]
> Hi, it would be nice to support IPv6 addresses to register ECS tasks in ALB target group (awsvpc mode). Right now, when I tried to enable IPv6 in the target group, ECS is trying to use ipv4 to register in the ipv6 target group:
> 
> ```
> service test-EcsService-fYJDJS3mLw5J failed to register targets in target-group test-tg  with (error The IP address '172.17.2.196' is not a valid IPv6 address)
> ```
> 
> The task itself has IPv6:
> 
> ```
> Private IP
> 172.17.2.196
> 
> IPv6 address
> 2a05:d018:1cee:b902:bd1e:854e:44a4:1f78
> ```

@relgames Were you able to resolve this? Hitting it right now, and I can't tell if it was just never implemented, or there's a switch I need to flip.

**Comment by relgames** [comment]
No, seems like it's still not supported.

**Comment by timzuiddam** [comment]
And now it's February and we are hit with this. Still no update?

**Comment by gregwebs** [comment]
There are many AWS services lacking in IPV6 support: https://docs.aws.amazon.com/vpc/latest/userguide/aws-ipv6-support.html
AWS is making an estimated billion dollars a year (revenue, but there's certainly some profit there) charging for IPV4 addresses. That money could be used to fund a roadmap with commitments for full IPV6 support for various services.

**Comment by maxgruebneraeroqual** [comment]
We're also being reamed by AWS switching to charge for IPv4 addresses, but not providing IPv6 functionality required to move away from IPv4.

Pretty transparent cash grab from AWS, considering they'll make back the cost of the IPv4 addresses in under two years.

It's lovely running into situations like:

I need public addresses for ECS containers, because AWS Secrets Manager requires public IP addresses (for some reason?).
I can't make my ECS containers IPv6 only (for some reason?), so all my containers now cost me an extra $50 a year, or $150 if I have a load balancer in front of them (which are also only dualstack and must have at least two IPv4 addresses themselves).

Allegedly this can be worked around by using PrivateLink, so we get to waste a bunch of time trying to get that working (and hoping that there's no extortionate charges involved in that too).

**Comment by jnawk** [comment]
> I need public addresses for ECS containers, because AWS Secrets Manager requires public IP addresses (for some reason?). I can't make my ECS containers IPv6 only (for some reason?), so all my containers now cost me an extra $50 a year, or $150 if I have a load balancer in front of them (which are also only dualstack and must have at least two IPv4 addresses themselves).
> 
> Allegedly this can be worked around by using PrivateLink, so we get to waste a bunch of time trying to get that working (and hoping that there's no extortionate charges involved in that too).

While this is a pretty shit move by AWS, it does sound like you can save a bit of money. 
If you've got a load balancer, then you don't need your containers to have public IP addresses - chuck them in private subnets.  You'll need a NAT gateway or NAT instance to provide them with outbound internet access to access the secrets manager endpoint, OR, you can create an interface endpoint for secrets manager in a private subnet, which will give your containers access to secrets manager without needing outbound internet access - there's a per hour charge associated, naturally - both with any NAT solution you use, and for the interface endpoint, but run the numbers, it might come to less than what you're otherwise paying. 

Of note, by running your containers in private subnets, you'll need to make sure they can access your container registry - if that's ECR, then you'll need an ECR interface endpoint, and a S3 gateway endpoint.   Same story with any other things your containers access.

Yeah, it might cost more to sort it out than you'd otherwise pay to AWS for IP addresses, so you might be better just paying that racket.

**Comment by Spacefish** [comment]
A first step would be to add Ipv6 support to ECR / dkr.ecr..
If i look at my ECR Registry, it´s CNAME pointing to a public network load balancer: `nlb3-21115c3760ec6ec9.elb.ap-east-1.amazonaws.com`
but this hostname only has an `A` record but no `AAAA` record.

Same thing for the layer download endpoint as well, `prod-ap-east-1-starport-layer-bucket.s3.ap-east-1.amazonaws.com` is a CNAME to `s3-r-w.ap-east-1.amazonaws.com` which has no `AAAA` record.

**Comment by klevo** [comment]
C'mon AWS this is embarrassing. Let's get that ECR IPv6 support in at least.

**Comment by mocknen** [comment]
I also tried, but still failed to specify ECS configured with IPv6 and Fargate as the target group for ALB, with the following error message.

> service <SERVICE_NAME> failed to register targets in target-group <TARGET_GROUP_NAME> with (error The IP address '<ENI_IPV4_ADDRESS>' is not a valid IPv6 address)

I am certain that the ENI configured for the task has an valid IPv6 address.
I have also tried configuring my account for `dualStackIPv6` as described in the [documentation](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/ecs-account-settings.html), but nothing has changed.

**Comment by TeoTN** [comment]
The whole issue could have been easily solved by AWS for low-cost tinkerers by offering up to 2-3 public ipv4 addresses in Free Tier, assignable to ALB, just like they offer one for an EC2 instance.

Given that barely anything at AWS at the moment supports ipv6, it's a ludicrous move. It's no customer centricity just purely squeezing as much money from people as possible, for no value added.

**Comment by gauravgoyal591** [comment]
Due to the AWS team charging for IPv4 and not being ready for IPv6 connectivity, please refrain from charging for IPv4 until your architecture is ready.

**Comment by TJNII** [comment]
Is this expected?

```
$ dig AAAA ecs.us-east-2.amazonaws.com

; <<>> DiG 9.11.4-P2-RedHat-9.11.4-26.P2.amzn2.13.5 <<>> AAAA ecs.us-east-2.amazonaws.com
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 4379
;; flags: qr rd ra; QUERY: 1, ANSWER: 0, AUTHORITY: 1, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 4096
;; QUESTION SECTION:
;ecs.us-east-2.amazonaws.com.   IN      AAAA

;; AUTHORITY SECTION:
ecs.us-east-2.amazonaws.com. 59 IN      SOA     ns-1999.awsdns-57.co.uk. awsdns-hostmaster.amazon.com. 1 7200 900 1209600 60

;; Query time: 0 msec
;; SERVER: 10.192.0.2#53(10.192.0.2)
;; WHEN: Thu Feb 29 02:11:06 UTC 2024
;; MSG SIZE  rcvd: 170
```

IPv6 in my VPC is working, I was able to install dig into the ec2 instance running amazon/amzn2-ami-ecs-hvm-2.0.20240109-x86_64-ebs over IPv6.  When I search for "ECS IPv6" the top hit is a press release proudly proclaiming "Amazon ECS now supports Internet Protocol Version 6 (IPv6) in awsvpc networking mode" so I'd expect this to work.

**Comment by gauravgoyal591** [comment]
No it's not working from the IPV6 servers

Connect timeout on endpoint URL: "https://api.ecr.us-east-1.amazonaws.com/"
[image: birdeye-icon]
<https://birdeye.com/?utm_source=product_email&utm_medium=customers&utm_campaign=default_signature&utm_term=birdeye_icon>

*Gaurav Kumar Goyal*

Software DevOps Engineer
(797) 620-2418
birdeye.com
<https://birdeye.com/?utm_source=product_email&utm_medium=customers&utm_campaign=default_signature&utm_term=birdeye_dot_com>

**Comment by CGarces-Apser** [comment]
> Is this expected?
> 
> ```
> $ dig AAAA ecs.us-east-2.amazonaws.com
> 
> ; <<>> DiG 9.11.4-P2-RedHat-9.11.4-26.P2.amzn2.13.5 <<>> AAAA ecs.us-east-2.amazonaws.com
> ;; global options: +cmd
> ;; Got answer:
> ;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 4379
> ;; flags: qr rd ra; QUERY: 1, ANSWER: 0, AUTHORITY: 1, ADDITIONAL: 1
> 
> ;; OPT PSEUDOSECTION:
> ; EDNS: version: 0, flags:; udp: 4096
> ;; QUESTION SECTION:
> ;ecs.us-east-2.amazonaws.com.   IN      AAAA
> 
> ;; AUTHORITY SECTION:
> ecs.us-east-2.amazonaws.com. 59 IN      SOA     ns-1999.awsdns-57.co.uk. awsdns-hostmaster.amazon.com. 1 7200 900 1209600 60
> 
> ;; Query time: 0 msec
> ;; SERVER: 10.192.0.2#53(10.192.0.2)
> ;; WHEN: Thu Feb 29 02:11:06 UTC 2024
> ;; MSG SIZE  rcvd: 170
> ```
> 
> IPv6 in my VPC is working, I was able to install dig into the ec2 instance running amazon/amzn2-ami-ecs-hvm-2.0.20240109-x86_64-ebs over IPv6. When I search for "ECS IPv6" the top hit is a press release proudly proclaiming "Amazon ECS now supports Internet Protocol Version 6 (IPv6) in awsvpc networking mode" so I'd expect this to work.

ECS **partially** support IPv6 but ECR not, ECR is the big issue
https://docs.aws.amazon.com/vpc/latest/userguide/aws-ipv6-support.html

**Comment by pawelfraczyk** [comment]
> Hi, it would be nice to support IPv6 addresses to register ECS tasks in ALB target group (awsvpc mode). Right now, when I tried to enable IPv6 in the target group, ECS is trying to use ipv4 to register in the ipv6 target group:
> 
> ```
> service test-EcsService-fYJDJS3mLw5J failed to register targets in target-group test-tg  with (error The IP address '172.17.2.196' is not a valid IPv6 address)
> ```
> 
> The task itself has IPv6:
> 
> ```
> Private IP
> 172.17.2.196
> 
> IPv6 address
> 2a05:d018:1cee:b902:bd1e:854e:44a4:1f78
> ```

End of March getting the same error.

VPC enabled ipv6, all subnets with routing tables.
ALB in dual-stack mode
ECS enabled in region for whole account to use dual-stack mode
Target group with ipv6

ECS Fargate Task has both ipv4 & ipv6 and can't register as a target in TG because it tries ipv4....

Any solutions?

**Comment by Laurenfrost** [comment]
Hello! Any udpates here?

I got exactly the same error when I tried to register a ecs fargate task to an ipv6 target group.

```
service `test-service` failed to register targets in target-group `test-tg-tbt-v6` with (error The IP address '192.168.162.68' is not a valid IPv6 address)
```

**Comment by ajpauwels** [comment]
@pawelfraczyk @Laurenfrost 

There is unfortunately no solution here, it requires AWS changing things on the back-end to detect if a target group is IPv6-only and then registering IPv6 targets rather than IPv4 targets.

**Comment by victorvhpg** [comment]
Any update ? @tabern

**Comment by arieldossantos** [comment]
Any update? I still can't connect over ipv6

**Comment by CGarces** [comment]
> Hello! Any udpates here?
> 
> I got exactly the same error when I tried to register a ecs fargate task to an ipv6 target group.
> 
> ```
> service `test-service` failed to register targets in target-group `test-tg-tbt-v6` with (error The IP address '192.168.162.68' is not a valid IPv6 address)
> ```

hi @Laurenfrost 
This change from AWS fix your issue?
https://aws.amazon.com/blogs/networking-and-content-delivery/introducing-dual-stack-without-public-ipv4-application-load-balancer/

**Comment by yann-soubeyrand** [comment]
@CGarces not for us: we have a dualstack NLB and we’d like our ECS services to register with IPv6 target groups, which is still failing.

**Comment by shmcgough** [comment]
Its a complete joke that you charge for IPV4 and then don't provide a way to actually use IPV6. Amazon soaks up thousands of engineers and what are they all doing?

**Comment by CGarces** [comment]
>  what are they all doing?
Profit

**Comment by will-daniels-d1** [comment]
Just vote with your wallets people, we switched to GCP for our latest project pretty much as a direct result of AWS exploiting this, even though the actual financial impact is limited. I was more mad about wasting nearly 2 days switching everything over to IPv6 only to find it couldn't actually be done because ECR and something else didn't support IPv6. Unfortunately have to wait out a 3 year savings plan for the existing services. If they were up-front about the blockers, or in any way interested in solving them, that would be different.

**Comment by bgre033** [comment]
Hello, while this is rather annoying, a workaround using FCK NAT seems to work pretty well.

https://medium.com/@gregor_68881/transitioning-aws-services-to-ipv6-following-ipv4-charges-61bcd854e637

**Comment by campmesh** [comment]
Thank you for your continued feedback on IPv6 support for ECS. I want to share our plans to address these challenges.

We are planning comprehensive IPv6-only support for Amazon ECS that aims to enable:
- Native IPv6 support for both EC2 and Fargate launch types
- IPv6-based communication with core AWS services
- The ability to register ECS tasks with IPv6 target groups for both Application and Network Load Balancers
- IPv6-compatible service discovery and networking through Cloud Map, Service Connect, and VPC Lattice

This implementation would eliminate the need for NAT infrastructure and IPv4 address management overhead in IPv6-only subnets. For workloads that still need to interact with IPv4-only services, we're planning to support DNS64/NAT64 translation.

To address specific concerns raised in this thread:
- Bridge networking mode IPv6 support is not currently planned for the initial release
- We plan to enable native IPv6 image pulls from ECR without requiring NAT gateways

While we're actively working on this feature set, these plans are subject to change as we continue development.

**Comment by wesmontgomery** [comment]
Looks like the AI hype cycle is finally cooling and engineers are finally looking at this?

Keep the pressure on!

**Comment by wesmontgomery** [comment]
Hey @campmesh @jenmlinaws just checking in! It's been about a month now since the plans have been [announced](https://github.com/aws/containers-roadmap/issues/1340#issuecomment-2680134619). 

Very exciting to see that this feature request is now under "[Work In Progress](https://github.com/aws/containers-roadmap/issues?q=state%3Aopen%20label%3A%22Work%20in%20Progress%22)"!

Do we have a timeline on when this will be ready?

Still cannot pull images from ECR using IPv6 only tasks. It also looks like CloudWatch Logs do not work with IPv6 only, is ECS configured to use the [dualstack  service endpoint](https://docs.aws.amazon.com/general/latest/gr/cwl_region.html#cwl_region)?

**Comment by wesmontgomery** [comment]
Would you look at that!

https://aws.amazon.com/about-aws/whats-new/2025/05/amazon-ecr-support-ipv6/

**Comment by tophercullen** [comment]
@wesmontgomery Have you actually gotten it to work with private registries? I've tried EKS, ECS, and locally using the dual stack endpoints and it doesn't seem to actually work correctly. I just get auth errors doing basic image pulls.

**Comment by wesmontgomery** [comment]
I haven't tested fully yet. However, I do remember the last time I used a 3rd party container provider that supported IPv6 that I had to also configure my Fargate task to not use the built in Cloudwatch integration as it was using the IPv4 only endpoint. There may be another service in your workload that still requires IPv4 😕

**Comment by tophercullen** [comment]
@wesmontgomery AWS' ECR cred helper does not work with the new endpoints:

`level=error msg="Error parsing the serverURL" error="docker-credential-ecr-login can only be used with Amazon Elastic Container Registry." serverURL=<redacted>.dkr-ecr.<redacted>.on.aws`

Issue already open: https://github.com/awslabs/amazon-ecr-credential-helper/issues/967

**Comment by pioneer2k** [comment]
@wesmontgomery  I have similar problems. We use AWS ECS Fargate with one private container from ECR and one public container from public ECR. The dual stack endpoint for the public ECR is working fine. But the dual stack endpoint of the private ECR is giving me: "authorization failed: no basic auth credentials". I tried this as well manually from an EC2 instance inside the same VPC. I got the same error. And we are working with multiple dual stack endpoints of other services (SNS/SQS/S3) and they are all working well.
Does someone got the dual stack endpoints of ECR to work?

**Comment by tophercullen** [comment]
@pioneer2k Thats the exact behavior I got. Based on the error message, I'm assuming that Fargate is using the ECR credential helper (or similar) under the hood. This currently doesn't support the dual stack endpoints. An update has been merged into the master branch, but there hasn't been a new release yet. Once there is, then its up to the AWS team to get the version updated in the backend.

Similar case for all AWS services, like EC2 and EKS with their respective credential helper package. The only workaround i can think of for EC2 (or locally), is doing the old AWS CLI/Docker login method. Obviously, that won't work for Fargate (or other AWS hosted services). You're just waiting on AWS there to fix it.

**Comment by pioneer2k** [comment]
@tophercullen I already tried the old way on an EC2 machine from the same VPC with "aws ecr get-login-password" + "docker login" + "docker pull" and I get the same error message ("authorization failed: no basic auth credentials"). Maybe the mentioned helpers are not the only problem here.....

**Comment by tophercullen** [comment]
@pioneer2k I just tested that login/password method successfully. I would verify your config, IAM etc. and/or that the pull isn't trying to use the cred helper anyway.

**Comment by pioneer2k** [comment]
@tophercullen  you are right. I tested again and it worked. Must be something wrong on my side yesterday.

**Comment by TommiFore** [comment]
How about using CloudWatch Logs dual stack endpoint with Fargate service, is that possible? I also tried to configure it manually with environment variable AWS_ENDPOINT_URL_LOGS with value "logs.${AWS::Region}.api.aws" but with no luck, it's still using old IPv4-only endpoint through NAT.

**Comment by ekini** [comment]
I made a quick terraform module to workaround the authorization issue - it runs a Lambda that creates a secret with Docker auth data and updates it periodically. The ECS Task Definition then references the secret and can successfully pull images using ipv6.

https://github.com/koandata/terraform-aws-ecs-ecr-ipv6-workaround feel free to use it.

For logs I'm pretty sure you can use the fluentbit driver, which should be able to use the ipv6 endpoint - haven't tested it yet, though.

But yay! Finally I have a full web stack that doesn't have ANY public ipv4 addresses: Cloudfront now has VPC origins, which can connect to an internal ALB. ECS supports pulling images from ECR. It's also able to pull secrets from the Secrets Manager successfully too. Amazing!

**Comment by Spacefish** [comment]
@ekini cool idea with updating the secret periodically in the secret manager, tried that manually as well :)
Hope AWS just adds the new pattern to the ECS platform to automatically do the auth.

**Comment by TommiFore** [comment]
> For logs I'm pretty sure you can use the fluentbit driver, which should be able to use the ipv6 endpoint - haven't tested it yet, though.

I don't know much about log drivers but according to [CFN docs](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-ecs-taskdefinition-logconfiguration.html#cfn-ecs-taskdefinition-logconfiguration-logdriver) only drivers allowed with Fargate are "awslogs", "splunk" and "awsfirelens".

You have to mean that "awsfirelens", which uses Fluent Bit sidecar or something like that behind the scenes. Am I right?

**Comment by ekini** [comment]
I've added the "awsfirelens" example - https://github.com/koandata/terraform-aws-ecs-ecr-ipv6-workaround/commit/0aa2010b891e65751810b8e6dcdaee920cec19c3

**Comment by TommiFore** [comment]
> I've added the "awsfirelens" example - [koandata/terraform-aws-ecs-ecr-ipv6-workaround@0aa2010](https://github.com/koandata/terraform-aws-ecs-ecr-ipv6-workaround/commit/0aa2010b891e65751810b8e6dcdaee920cec19c3)

I managed to test this and it almost worked nicely. Unfortunately Fluent Bit does not use IPv6 reliably and mainly it tries to use IPv4 for communication to dual stack endpoint of CloudWatch Logs.

Small portion of logs are delivered properly, but big portion is missing and there is "rejected" entries in VPC flow logs with destinations of this endpoint's IPv4 addresses. That is achieved by blocking non-intra-VPC IPv4 traffic in security group. Obviously it will be using NAT if it's allowed.

Do you know some configuration options which I didn't find to force Fluent Bit to use only IPv6?

**Comment by TommiFore** [comment]
I did the most obvious solution and created CloudFront distribution only with AAAA record and now the Fluent Bit works properly. If someone is also interested in, I should mention that ECS pulls the secrets from Secrets Manager also over IPv4 and not IPv6 although the Secrets Manager has normal dual stack endpoint. You can move secrets into environment variables and resolve the values with CloudFormation already. Maybe you want to check ECS permissions to read those variables by your users...

Probably every case is different but now it works completely without public IPv4 for our use cases and I mean no need for NAT and you can only allow intra-VPC IPv4 egress traffic. Must to say that ECS is still pretty far from plug & play experience with IPv6-only usage but it's doable...

**Comment by ekini** [comment]
AWS-supported Fluentbit is very old - it's based on 1.9 version which was EOL in 2022 - https://github.com/aws/aws-for-fluent-bit/issues/885#issuecomment-2787814869

You could try using a recent upstream fluentbit 4.x, but I'm not sure how it will work with the AWS Firelens integration. Newer fluentbit versions have "prefer ipv6" config option.

On the other hand, some improvements have been made to https://github.com/aws/amazon-ecs-agent/releases/tag/v1.93.0 so presumably it's used in Fargate too and will be updated at some point? So that the normal awslogs driver will just work. Hopefully.

**Comment by TommiFore** [comment]
CloudFront proxying is okay for now, I think. There is configuration option "awslogs-endpoint" in ECS agent codes, but at the moment CloudFormation complains with error "Invalid request provided: Create TaskDefinition: Log driver awslogs disallows options: awslogs-endpoint" but maybe in some day it just works... Thanks for help.

**Comment by nemosupremo** [comment]
At around August 11, 2025 at 15:37 (UTC-7:00) our IPv6 only ECS tasks stopped working entirely. We've been using the [terraform-aws-ecs-ecr-ipv6-workaround](https://github.com/koandata/terraform-aws-ecs-ecr-ipv6-workaround) workaround, as well as the `ecr-public.aws.com` images.

Suddenly those tasks fails with:

```
ResourceInitializationError: unable to pull secrets or registry auth: The task cannot pull registry auth from Amazon ECR: There is a connection issue between the task and Amazon ECR. Check your task network configuration. operation error ECR: GetAuthorizationToken, exceeded maximum number of attempts, 3, https response error StatusCode: 0, RequestID: , request send failed, Post "https://api.ecr.us-east-2.amazonaws.com/": dial tcp 3.17.137.51:443: i/o timeout
```

It's unclear if its the public ecr repo that is causing problems or the `dkr-ecr.us-east-2.on.aws` repos. Regardless, it looks like ECS is trying to make a request to non-dual stack endpoint.

**Comment by ekini** [comment]
It's broken for me too and AWS support is unhelpful. Haven't found a workaround yet.

**Comment by ekini** [comment]
Found a workaround.

I used `{account_number}.dkr-ecr.ap-southeast-2.on.aws` with my secret workaround and `ecr-public.aws.com` to firelens, which had been working since May until a few days ago.

The error above says the task tries to run `GetAuthorizationToken` (for `{account_number}.dkr-ecr.ap-southeast-2.on.aws`) and goes to `api.ecr.us-east-2.amazonaws.com` to do that. Funnily enough, `GetAuthorizationToken` call isn't even needed here as the secret workaround provides ECS with user/password to be able get access to `{account_number}.dkr-ecr.ap-southeast-2.on.aws`.

My guess is someone tried to make IAM authentication work with `{account_number}.dkr-ecr.ap-southeast-2.on.aws` and **vibe-coded** it to go to the ipv4-only endpoint `api.ecr.us-east-2.amazonaws.com`. Shame.

The solution is to make a bare Cloudfront distribution to proxy to `{account_number}.dkr-ecr.ap-southeast-2.on.aws` so that ECS doesn't try to call `GetAuthorizationToken`. Use version 0.0.4 of https://github.com/koandata/terraform-aws-ecs-ecr-ipv6-workaround

Then use it like this in terraform:
```hcl

  {
    name = "app"
    // ipv6 version
    image = "${module.ecs_ipv6.ecr_proxy_domain}/${local.name}:latest"
    repositoryCredentials = {
      credentialsParameter = module.ecs_ipv6.secret_arn
    }
  ...
  }
```

**Comment by Spacefish** [comment]
Oh so they tried to fix it, but broke it in the process as `api.ecr.us-east-2.amazonaws.com` doesn´t have an AAAA record.
But there is a dual stack enpoint for that API `ecr.us-east-2.api.aws` probably they just fear changing the default to this endpoint, as this could break some workloads with a setup ipv6 but broken routing (potentially)..

Maybe they need to make it configurable / default to the dual-stack endpoint only if no ipv4 is configured or something like that?... Hard to migrate i guess with these "old" v4 only and "new" dual stack endpoints for everything..

The AWS CLI btw (and probably all the libs) default to the ipv4 only endoints b.t.w. you have to explicitly add `--endpoint-url https://ecr.us-east-2.api.aws` to make it use the new endpoints..

But on the plus side, using the new `[accountnumber].dkr-ecr.[region].on.aws` works now if your task still has an public ipv4! Nice 👍🏻.

Migrated all my ECS Tasks to the new Endpoints now!

**Comment by BwL1289** [comment]
There an ETA to land this?

**Comment by wesmontgomery** [comment]
It think it's really official now! https://aws.amazon.com/about-aws/whats-new/2025/09/amazon-ecs-ipv6-only-support/

**Comment by TommiFore** [comment]
I tested it today and noticed there are not any logs through Fluent Bit + CloudFront workaround. Turns out that IAM auth is now signing also with Host header, so origin request policy must be "AllViewer" and then it works again.

In dual stack subnets all requests are still using IPv4, but maybe IPv6-only subnets will work, who knows...

**Comment by tophercullen** [comment]
@wesmontgomery Despite the blog post saying "Integrate with AWS PrivateLink  and Amazon Virtual Private Cloud endpoints using IPv6", it does not. ECR PrivateLink cannot be configured correctly for dual stack or ipv6, and clearly and demonstrably still does not work with IPv6.

I just tried to provision an ECS service in an ipv6 only subnet and still got this: "PrivateLink is not supported when connecting to ECR using the dualstack domain". 

Dig from that VPC shows PrivateLink still answers for the new dual stack domains (despite it not ever actually supporting dual stack) and provides no AAAA record (the endpoint interfaces having Ipv6 addresses themselves). 

So mixed workloads are off the table, and migrations may be tricky in some cases.

**Comment by jnawk** [comment]
You can host the necessary DNS zone yourself with a bit of extra work.    That'll solve the no AAAA thing. Not sure if it'll complete the puzzle, but worth a try.

**Comment by ekini** [comment]
I have a support ticket with AWS, and the awslogs driver is expected to just work via ipv6 at some point on Fargate. It might already work on EC2-backed ECS using the last versions of ecs-agent.

No update on the ECR authentication side of things.

**Comment by Spacefish** [comment]
> No update on the ECR authentication side of things.

Oh that´s interesting... So it works if you host you own registry or use dockerhub but not with their managed service lol :(

**Comment by ekini** [comment]
Just tested that both ECR auth and awslogs work in **IPv6-only** subnets.

**Comment by Rush-iam** [comment]
My AutoScaling group spawns an EC2 instance in an IPv6-only network, but it does not appear in the ECS as an available instance. It works if I assign a public IPv4, but I am getting "agent disconnected" if I remove it.

Should the feature added today allow ECS on EC2 without a public IPv4?

**Comment by nemosupremo** [comment]
I'm, bizarrely, now seeing authentication errors when trying to access public images using ecr's public gallery with the dual stack endpoint:

```
CannotPullContainerError: pull image manifest has been retried 1 time(s): failed to resolve ref ecr-public.aws.com/aws-cli/aws-cli:2.4.6@sha256:d123012561ce8a323782d891d79ef81791f49f964692e42b4e00e89d2c5669a3: unexpected status from HEAD request to https://ecr-public.aws.com/v2/aws-cli/aws-cli/blobs/sha256:d123012561ce8a323782d891d79ef81791f49f964692e42b4e00e89d2c5669a3: 401 Unauthorized
```

**Comment by nikitadahiya** [comment]
> My AutoScaling group spawns an EC2 instance in an IPv6-only network, but it does not appear in the ECS as an available instance. It works if I assign a public IPv4, but I am getting "agent disconnected" if I remove it.
> 
> Should the feature added today allow ECS on EC2 without a public IPv4?

With the new IPv6-only support, ECS container instances on EC2 don’t need a public IPv4 address. One key requirement is that each instance must be assigned a primary IPv6 address at launch.

If you’re using an Auto Scaling Group, check that the setting to assign a primary IPv6 address is turned on. Otherwise, your instances may come up without one, and the ECS agent will fail to connect (showing “agent disconnected”).

From the [ECS documentation](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/task-networking.html):

`“When you launch a container instance for running tasks in an IPv6-only configuration, you must set a primary IPv6 address for the instance by using the --enable-primary-ipv6 EC2 parameter.”`

So if you’re hitting this issue, make sure your instances are always being launched with a primary IPv6 address.

**Comment by nikitadahiya** [comment]
> I'm, bizarrely, now seeing authentication errors when trying to access public images using ecr's public gallery with the dual stack endpoint:
> 
> ```
> CannotPullContainerError: pull image manifest has been retried 1 time(s): failed to resolve ref ecr-public.aws.com/aws-cli/aws-cli:2.4.6@sha256:d123012561ce8a323782d891d79ef81791f49f964692e42b4e00e89d2c5669a3: unexpected status from HEAD request to https://ecr-public.aws.com/v2/aws-cli/aws-cli/blobs/sha256:d123012561ce8a323782d891d79ef81791f49f964692e42b4e00e89d2c5669a3: 401 Unauthorized
> ```

Can you share more details of your task setup, is it a Fargate task or a EC2 task, if it’s EC2, what is the networking mode being used, and what’s the IP address type of the instance?

**Comment by ekini** [comment]
> > I'm, bizarrely, now seeing authentication errors when trying to access public images using ecr's public gallery with the dual stack endpoint:
> > ```
> > CannotPullContainerError: pull image manifest has been retried 1 time(s): failed to resolve ref ecr-public.aws.com/aws-cli/aws-cli:2.4.6@sha256:d123012561ce8a323782d891d79ef81791f49f964692e42b4e00e89d2c5669a3: unexpected status from HEAD request to https://ecr-public.aws.com/v2/aws-cli/aws-cli/blobs/sha256:d123012561ce8a323782d891d79ef81791f49f964692e42b4e00e89d2c5669a3: 401 Unauthorized
> > ```
> 
> Can you share more details of your task setup, is it a Fargate task or a EC2 task, if it’s EC2, what is the networking mode being used, and what’s the IP address type of the instance?

I experience that too in a dualstack network. AWS Fargate, so awsvpc, tasks have both local ipv4 and ipv6 addresses. 

I wouldn't close this issue as it still persists for dualstack networks. Only works in ipv6 networks, which is limiting.
Tested on Fargate: awslogs does **not** work in a dualstack network. ECR does **not** work in a dualstack network.

**Comment by Rush-iam** [comment]
> > My AutoScaling group spawns an EC2 instance in an IPv6-only network, but it does not appear in the ECS as an available instance. It works if I assign a public IPv4, but I am getting "agent disconnected" if I remove it.
> > Should the feature added today allow ECS on EC2 without a public IPv4?
> 
> With the new IPv6-only support, ECS container instances on EC2 don’t need a public IPv4 address. One key requirement is that each instance must be assigned a primary IPv6 address at launch.

Thank you, @nikitadahiya. It turned out the reason was in the agent version - even though the "Update agent" action told me the instance is running the latest version, it was 1.91.2:

<img width="1038" height="81" alt="Image" src="https://github.com/user-attachments/assets/ee4050f6-380a-4b47-a985-01c97e52a205" />

Replacing the AMI image with the latest available AL2023 in the Autoscaling Group Launch Template solved the issue, and the agent is now 1.99.1 (matches min required for the feature), and IPv6-only EC2 + ECS now works ✌️

<img width="1036" height="77" alt="Image" src="https://github.com/user-attachments/assets/2f5daf8b-6327-44a3-8e10-c308a70ac24b" />

**Comment by TommiFore** [comment]
Is there any plans to support IPv6 properly in dual stack subnets?

**Comment by Rush-iam** [comment]
@TommiFore for EC2, to make it work in dual-stack subnets, I needed to add `echo ECS_INSTANCE_IP_COMPATIBILITY=ipv6 >> /etc/ecs/ecs.config` to UserData in the EC2 launch template

You may also need to configure Docker to prefer IPv6 by creating `/etc/docker/daemon.json` with the following content + restarting the Docker service:
```
{
  "ipv6": true,
  "fixed-cidr-v6": "2001:db8:1::/64"
}
```

(and also make sure that image ECR endpoints in the ECS task definition are ipv6-compatible, not the ones ECR gives by default)

**Comment by TommiFore** [comment]
That's great but we are using Fargate so any of these options are not configurable, right? Sorry, I forgot to mention it...

**Comment by chriscatuk** [comment]
with ECS Fargate, ipv6 only, egress only. In a VPC with both API and DKR VPC endpoints. in eu-west-3

one week ago, I moved to Docker Hub private repo for now because I couldn't make ECR work on ipv6-only with egress only. 

now that I discover this issue I moved back to ECR to share all the error messages. Surprise: it works!

somehow the ipv6-only started working in the past few days.

thanks!



At first I thought I found the way because I stopped the "timeout" by replacing
```
-  {account_id}.dkr.ecr.eu-west-3.amazonaws.com/{image_name}:{image_tag}
+  {account_id}.dkr-ecr.eu-west-3.on.aws/{image_name}:{image_tag}
```

but instead I got `405 Method Not Allowed`


now I directly went back to 
```
{account_id}.dkr.ecr.eu-west-3.amazonaws.com/{image_name}:{image_tag}
```

and it just works



Now I don't understand why we are forced to use ECR enpoints, it's 0.50 per AZ per day that are not necessary for most projects. It just timeouts

```
 request send failed, Post "https://api.ecr.eu-west-3.amazonaws.com/": dial tcp 35.180.245.30:443: i/o timeout.
```
